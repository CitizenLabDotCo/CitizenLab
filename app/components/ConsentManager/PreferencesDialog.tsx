import React, { PureComponent } from 'react';

// Components
import CategoryCard from './CategoryCard';

// Typing
import { IDestination } from './';

// Styling
import styled from 'styled-components';
import { fontSizes, media } from 'utils/styleUtils';

export const ContentContainer = styled.div`
  padding: 30px;
  background: white;

  h1 {
    font-size: ${fontSizes.medium}px;
  }

  h2 {
    font-size: ${fontSizes.large}px;
  }

  ${media.smallerThanMaxTablet`
    margin: 0;
    padding: 20px;
  `}
`;

interface Props {
  onChange: (category, value) => void;
  categoryDestinations: {
    analytics: IDestination[];
    advertising: IDestination[];
    functional: IDestination[];
  };
  analytics: boolean | null;
  advertising: boolean | null;
  functional: boolean | null;
}

export default class PreferencesDialog extends PureComponent<Props> {
  static displayName = 'PreferencesDialog';

  handleChange = (event) => {
    event.preventDefault();
    this.props.onChange(event.target.name, event.target.value === 'true');
  }

  render() {
    const {
      categoryDestinations,
      functional,
      advertising,
      analytics,
    } = this.props;
    const checkCategories = { analytics, advertising, functional };
    return (
      <ContentContainer id="e2e-preference-dialog">
        {Object.keys(categoryDestinations).map((category) => {
          if (categoryDestinations[category].length > 0) {
            return (
              <CategoryCard
                key={category}
                category={category}
                destinations={categoryDestinations[category]}
                checked={checkCategories[category]}
                handleChange={this.handleChange}
              />
            );
          }
          return;
        })}
      </ContentContainer>
    );
  }
}
