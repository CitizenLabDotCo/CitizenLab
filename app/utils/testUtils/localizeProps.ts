// i18n
import { getLocalized } from 'utils/i18n';
import { InjectedLocalized } from 'utils/localize';

// Typing
import { Multiloc } from 'typings';

// A mock object to replace injectLocalize function : pass this to the child of injectLocalize as a prop
// ex : injectLocalize(TestedComponent) =>
// <TestedComponent {...localizeProps}>
export const localizeProps = {
  localize: (multiloc: Multiloc) => {
    return getLocalized(multiloc, 'en', ['en', 'fr-BE']);
  },
  locale: 'en',
  tenantLocales: ['en', 'fr-BE']
} as InjectedLocalized;
