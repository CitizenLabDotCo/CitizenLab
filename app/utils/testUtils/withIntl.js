import { IntlProvider, intlShape } from "react-intl";
import { shallow, mount } from "enzyme";
import React from 'react';

// Create IntlProvider to retrieve React Intl context
const intlProvider = new IntlProvider(
  {
    locale: "en",
    messages: {
      message1: "Hello world"
    }
  },
  {}
);
const { intl } = intlProvider.getChildContext();

// `intl` prop is required when using injectIntl HOC
const nodeWithIntlProp = node => React.cloneElement(node, { intl });

// shallow() with React Intl context
export const shallowWithIntl = (node, { context, ...options } = {}) => {
  return shallow(nodeWithIntlProp(node), {
    ...options,
    context: {
      ...context,
      intl
    }
  });
};

// mount() with React Intl context
export const mountWithIntl = (
  node,
  { context, childContextTypes, ...options } = {}
) => {
  return mount(nodeWithIntlProp(node), {
    ...options,
    context: {
      ...context,
      intl
    },
    childContextTypes: {
      intl: intlShape,
      ...childContextTypes
    }
  });
};
