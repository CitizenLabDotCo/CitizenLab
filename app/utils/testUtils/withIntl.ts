import { IntlProvider, intlShape, InjectedIntlProps } from 'react-intl';
import { shallow, mount, ShallowRendererProps, ShallowWrapper } from 'enzyme';
import React from 'react';

// Create IntlProvider to retrieve React Intl context
const intlProvider = new IntlProvider({
  locale: 'en',
}, {});

// You customize the intl object here:
const { intl: originalIntl } = intlProvider.getChildContext() as InjectedIntlProps;
const intl = {
  ...originalIntl,
  formatMessage: ({ id, defaultMessage }, values?) =>
    originalIntl.formatMessage({
      id,
      defaultMessage: defaultMessage || id
    }, { tenantName: 'The Test', orgName: 'Test Town', orgType: 'testing', ...values || {} }),
};

const nodeWithIntlProp = (node) => React.cloneElement(node, { intl });

// shallow() with React Intl context
export function shallowWithIntl<C extends React.Component, P = C['props'], S = C['state']>(node: React.ReactElement<P>, additional: ShallowRendererProps = {}) {
 const { context, ...options } = additional;
  return shallow(nodeWithIntlProp(node), {
    ...options,
    context: {
      ...context,
      intl
    }
  }) as ShallowWrapper<P & InjectedIntlProps, S, C>;
}

// mount() with React Intl context
export const mountWithIntl = (
  node,
  { context, childContextTypes, ...options } = {}
) => {
  return mount(nodeWithIntlProp(node), {
    ...options,
    context: {
      ...context,
      intl
    },
    childContextTypes: {
      intl: intlShape,
      ...childContextTypes
    }
  });
};
