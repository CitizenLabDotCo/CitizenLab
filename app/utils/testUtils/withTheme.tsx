import React from 'react';
import { ThemeProvider } from 'styled-components';
import { colors, fontSizes, stylingConsts } from 'utils/styleUtils';
import { shallow, mount } from 'enzyme';

export const theme = {
  colors,
  fontSizes,
  colorMain: '#ef0071',
  colorText: '#e68f51',
  colorSecondary: '#f76901',
  menuStyle: 'light',
  menuHeight: 74,
  mobileMenuHeight: 72,
  mobileTopBarHeight: 66,
  maxPageWidth: 952,
  ...stylingConsts
};

// export function shallowWithTheme<C extends Component, P = C['props'], S = C['state']>(tree: ReactElement<P>, options?: ShallowRendererProps) {
//   const context = shallow(<ThemeProvider theme={theme} />)
//     .instance()
//     .getChildContext();
//   return shallow(tree, {
//     ...options,
//     context: {
//       ...context,
//       intl
//     }
//   }) as ShallowWrapper<P, S, C>;
// }
//
export const shallowWithTheme = (tree) => {
  const context = (shallow(<ThemeProvider theme={theme} />).instance() as any).getChildContext();
  return shallow(tree, { context });
};

export const mountWithTheme = (tree) => {
  const context = shallow(<ThemeProvider theme={theme} />)
    .instance()
    // @ts-ignore: TODO type this well
    .getChildContext();
  return mount(tree, {
    context,
    childContextTypes: ThemeProvider.childContextTypes,
  });
};
//
// export const mountWithTheme = (children: any) => mount(<ThemeProvider theme={theme}>{children}</ThemeProvider>);
//
// export const shallowWithTheme = (children: any) => shallow(<ThemeProvider theme={theme}>{children}</ThemeProvider>);
