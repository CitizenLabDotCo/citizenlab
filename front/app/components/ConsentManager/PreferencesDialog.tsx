import React from 'react';

// Components
import CategoryCard from './CategoryCard';

// Typing
import { CategorizedDestinations, IPreferences } from '.';

// Styling
import styled from 'styled-components';
import { fontSizes, media } from 'utils/styleUtils';

// services
import { TCategory } from './destinations';

export const ContentContainer = styled.div`
  padding: 30px;
  background: white;

  h1 {
    font-size: ${fontSizes.m}px;
  }

  h2 {
    font-size: ${fontSizes.l}px;
  }

  ${media.phone`
    margin: 0;
    padding: 20px;
  `}
`;

interface Props {
  onChange: (category: TCategory, value) => void;
  categoryDestinations: CategorizedDestinations;
  preferences: IPreferences;
}

const doNothing = () => () => {};

const PreferencesDialog = ({
  categoryDestinations,
  preferences,
  onChange,
}: Props) => {
  const handleChange = (category: TCategory, value: boolean) => (_event) => {
    onChange(category, value);
  };

  return (
    <ContentContainer id="e2e-preference-dialog">
      {Object.keys(categoryDestinations)
        .filter((category: TCategory) => {
          return categoryDestinations[category].length > 0;
        })
        .map((category: TCategory) => {
          return (
            <CategoryCard
              key={category}
              category={category}
              destinations={categoryDestinations[category]}
              checked={!!preferences[category]}
              handleChange={handleChange}
            />
          );
        })}
      <CategoryCard
        key={'required'}
        category={'required'}
        destinations={[]}
        checked={true}
        handleChange={doNothing}
        disableUncheck={true}
      />
    </ContentContainer>
  );
};

export default PreferencesDialog;
