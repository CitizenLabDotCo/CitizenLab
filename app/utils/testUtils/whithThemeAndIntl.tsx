// To render components with both intl and theme
// Never used yet, not sure how well it works
import React from 'react';
import { ThemeProvider } from 'styled-components';
import { colors, fontSizes } from '../styleUtils';

import { shallow } from 'enzyme';

import { shallowWithIntl, mountWithIntl } from 'utils/testUtils/withIntl';
const theme = {
  colors,
  fontSizes,
  colorMain: '#ef0071',
  menuStyle: 'light',
  menuHeight: 74,
  mobileMenuHeight: 72,
  mobileTopBarHeight: 66,
  maxPageWidth: 952,
};

export const shallowWithThemeIntl = (tree) => {
  const context = shallow(<ThemeProvider theme={theme} />)
    .instance()
    .getChildContext();
  return shallowWithIntl(tree, { context });
};
export const mountWithThemeIntl = (tree) => {
  const context = shallow(<ThemeProvider theme={theme} />)
    .instance()
    .getChildContext();
  return mountWithIntl(tree, {
    context,
    childContextTypes: ThemeProvider.childContextTypes,
  });
};
