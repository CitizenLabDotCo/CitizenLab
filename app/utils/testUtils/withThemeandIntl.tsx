import React from 'react';
import { ThemeProvider } from 'styled-components';
import { colors, fontSizes } from '../styleUtils';

import { shallow, mount } from 'enzyme';

import { shallowWithIntl, mountWithIntl } from 'utils/testUtils/withIntl';

const theme = {
  colors,
  fontSizes,
  colorMain: '#ef0071',
  menuStyle: 'light',
  menuHeight: 74,
  mobileMenuHeight: 72,
  mobileTopBarHeight: 66,
  maxPageWidth: 952,
};

export const shallowWithHoCs = (tree) => {
  const context = shallow(<ThemeProvider theme={theme} />)
    .instance()
    .getChildContext();
  return shallowWithIntl(tree, { context });
};
export const mountWithHoCs = (tree) => {
  const context = shallow(<ThemeProvider theme={theme} />)
    .instance()
    .getChildContext();
  return mountWithIntl(tree, {
    context,
    childContextTypes: ThemeProvider.childContextTypes,
  });
};
