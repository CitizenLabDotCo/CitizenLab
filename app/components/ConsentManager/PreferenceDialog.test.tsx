// libraries
import React from 'react';
import { shallow } from 'enzyme';
import { mountWithTheme } from 'utils/testUtils/withTheme';

import 'jest-styled-components';

// component to test
import PreferencesDialog from './PreferencesDialog';

// mock utilities
jest.mock('utils/cl-intl');

describe('<ConsentManager />', () => {
  let onCancel: Jest.Mock;
  let onSave: Jest.Mock;
  let onChange: Jest.Mock;
  const isConsentRequired = true;
  const implyConsentOnInteraction = false;

  const categoryDestinations = {
    analytics: [
      {
        name: 'Google Analytics',
        description: 'Google Analytics is the most popular marketing tool for the web. It’s free and provides a wide range of features. It’s especially good at measuring traffic sources and ad campaigns.',
        category: 'Analytics',
        website: 'http://google.com/analytics',
        id: 'Google Analytics',
      },
      {
        name: 'MarketingTool',
        description: 'MarketingTool is the most popular marketing tool for the web. It’s free and provides a wide range of features. It’s especially good at measuring traffic sources and ad campaigns.',
        category: 'Tag Managers',
        website: 'http://random.com/marketing',
        id: 'MarketingTool',
      },
    ],
    advertising: [
      {
        name: 'AdvertisingTool',
        description: 'Advertising BS',
        category: 'Advertising',
        website: 'http://random.com/advertising',
        id: 'AdvertisingTool',
      },
    ],
    functional: [
      {
        name: 'FunctionalTool',
        description: 'Actually might be handy',
        category: 'Security & Fraud',
        website: 'http://random.com/securitycookie',
        id: 'FunctionalTool',
      }
    ]
  };
  const preferences = {
    analytics: true,
    advertising: null,
    functional: null
  };

  beforeEach(() => {
    onChange = jest.fn();
    onSave = jest.fn();
    onCancel = jest.fn();
  });

  it('renders correctly when there are destinations', () => {
    const wrapper = shallow(
      <PreferencesDialog
        onCancel={onCancel}
        onSave={onSave}
        onChange={onChange}
        categoryDestinations={categoryDestinations}
        analytics={preferences.analytics}
        advertising={preferences.advertising}
        functional={preferences.functional}
      />);
    expect(wrapper).toMatchSnapshot();
  });
  it('renders correctly when there are no destinations', () => {
    const wrapper = shallow(
      <PreferencesDialog
        onCancel={onCancel}
        onSave={onSave}
        onChange={onChange}
        categoryDestinations={{
          analytics: [],
          advertising: [],
          functional: []
        }}
        analytics={preferences.analytics}
        advertising={preferences.advertising}
        functional={preferences.functional}
      />);
    expect(wrapper).toMatchSnapshot();
  });
  it('has a disabled save button when some field is empty', () => {
    const wrapper = shallow(
      <PreferencesDialog
        onCancel={onCancel}
        onSave={onSave}
        onChange={onChange}
        categoryDestinations={categoryDestinations}
        analytics={preferences.analytics}
        advertising={preferences.advertising}
        functional={preferences.functional}
      />);
    expect(wrapper.find('.integration-save').prop('disabled')).toBeTruthy();
  });
  it('has an enabled save button when no field is empty', () => {
    const wrapper = shallow(
      <PreferencesDialog
        onCancel={onCancel}
        onSave={onSave}
        onChange={onChange}
        categoryDestinations={categoryDestinations}
        analytics={true}
        advertising={false}
        functional={true}
      />);
    expect(wrapper.find('.integration-save').prop('disabled')).toBeFalsy();
  });
});
