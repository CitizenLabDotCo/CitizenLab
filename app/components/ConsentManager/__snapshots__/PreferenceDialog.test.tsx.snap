// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConsentManager /> renders correctly when there are destinations 1`] = `
<PreferencesDialog__ContentContainer
  aria-modal={true}
  role="dialog"
>
  <FormattedMessage
    defaultMessage="Your cookie preferences"
    id="app.components.ConsentManager.Modal.PreferencesDialog.title"
    tagName="h1"
  />
  <PreferencesDialog__Scroll>
    <CategoryCard
      category="analytics"
      checked={true}
      destinations={
        Array [
          Object {
            "category": "Analytics",
            "description": "Google Analytics is the most popular marketing tool for the web. It’s free and provides a wide range of features. It’s especially good at measuring traffic sources and ad campaigns.",
            "id": "Google Analytics",
            "name": "Google Analytics",
            "website": "http://google.com/analytics",
          },
          Object {
            "category": "Tag Managers",
            "description": "MarketingTool is the most popular marketing tool for the web. It’s free and provides a wide range of features. It’s especially good at measuring traffic sources and ad campaigns.",
            "id": "MarketingTool",
            "name": "MarketingTool",
            "website": "http://random.com/marketing",
          },
        ]
      }
      handleChange={[Function]}
      key="analytics"
    />
    <CategoryCard
      category="advertising"
      checked={null}
      destinations={
        Array [
          Object {
            "category": "Advertising",
            "description": "Advertising BS",
            "id": "AdvertisingTool",
            "name": "AdvertisingTool",
            "website": "http://random.com/advertising",
          },
        ]
      }
      handleChange={[Function]}
      key="advertising"
    />
    <CategoryCard
      category="functional"
      checked={null}
      destinations={
        Array [
          Object {
            "category": "Security & Fraud",
            "description": "Actually might be handy",
            "id": "FunctionalTool",
            "name": "FunctionalTool",
            "website": "http://random.com/securitycookie",
          },
        ]
      }
      handleChange={[Function]}
      key="functional"
    />
  </PreferencesDialog__Scroll>
  <PreferencesDialog__Spacer />
  <PreferencesDialog__ButtonContainer>
    <WithTheme(Button)
      className="integration-cancel"
      onClick={[MockFunction]}
      style="primary-inverse"
    >
      <FormattedMessage
        defaultMessage="Cancel"
        id="app.components.ConsentManager.Modal.PreferencesDialog.cancel"
      />
    </WithTheme(Button)>
    <WithTheme(Button)
      className="integration-save"
      disabled={true}
      onClick={[Function]}
      style="primary"
    >
      <FormattedMessage
        defaultMessage="Save"
        id="app.components.ConsentManager.Modal.PreferencesDialog.save"
      />
    </WithTheme(Button)>
  </PreferencesDialog__ButtonContainer>
</PreferencesDialog__ContentContainer>
`;

exports[`<ConsentManager /> renders correctly when there are no destinations 1`] = `
<PreferencesDialog__ContentContainer
  aria-modal={true}
  role="dialog"
>
  <FormattedMessage
    defaultMessage="Your cookie preferences"
    id="app.components.ConsentManager.Modal.PreferencesDialog.title"
    tagName="h1"
  />
  <PreferencesDialog__Scroll />
  <PreferencesDialog__Spacer />
  <PreferencesDialog__ButtonContainer>
    <WithTheme(Button)
      className="integration-cancel"
      onClick={[MockFunction]}
      style="primary-inverse"
    >
      <FormattedMessage
        defaultMessage="Cancel"
        id="app.components.ConsentManager.Modal.PreferencesDialog.cancel"
      />
    </WithTheme(Button)>
    <WithTheme(Button)
      className="integration-save"
      disabled={false}
      onClick={[Function]}
      style="primary"
    >
      <FormattedMessage
        defaultMessage="Save"
        id="app.components.ConsentManager.Modal.PreferencesDialog.save"
      />
    </WithTheme(Button)>
  </PreferencesDialog__ButtonContainer>
</PreferencesDialog__ContentContainer>
`;
